### PR TITLE
AirfRANS: T_max=5 + grad-clip=1.0 + WD=1e-2 compound golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The two best AirfRANS results used different sub-configurations of the golden compound:

- **PR #2732** (kohaku, **best: 0.01271**): T_max=5 + grad-clip=1.0 + WD=**1e-4** (default)
- **PR #2709** (fern, 0.01323): T_max=**10** + grad-clip=1.0 + WD=1e-2

Neither has combined T_max=5 (more cosine cycles, deeper basin exploration) with WD=1e-2 (clean regularization enabled by grad-clip). Combining both should compound both improvements — this is the highest priority experiment in the AirfRANS pipeline.

**Why this should work:** T_max=5 produces 8 full cosine cycles in 41 epochs, enabling more frequent phase-transition opportunities. WD=1e-2 provides clean L2 regularization without gradient explosions at cosine peaks (grad-clip=1.0 prevents those spikes). Both effects are largely orthogonal — T_max controls exploration cadence, WD controls the loss landscape geometry.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999 \
  --wandb-group airfrans-tmax5-compounds \
  --wandb-name "kohaku/airfrans-tmax5-gc1-wd1e2"
```

**MANDATORY flags:** `--epochs 999` and `--no-use-ema` (both required for all AirfRANS experiments).

**What to report:**
- Best `val_primary/surface_mse` and which epoch it occurred at
- Whether the best epoch falls at a cosine trough or peak (T_max=5 means troughs at epochs ~5, 10, 15, 20, 25, 30, 35, 40)
- The W&B run ID
- Training dynamics: did a phase transition occur? At what epoch?

## Baseline

- **Primary metric:** `val_primary/surface_mse` (lower is better)
- **Current best:** 0.01271 (PR #2732, kohaku — T_max=5 + grad-clip=1.0 + WD=1e-4 default)
- **Baseline W&B run:** uh7fchiy
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --no-use-ema --enable-fourier --epochs 999`

**Prior results for context:**

| Config | val_primary/surface_mse |
|--------|------------------------|
| T_max=5 + gc=1.0 + WD=1e-4 (PR #2732) | 0.01271 ← BEAT THIS |
| T_max=10 + gc=1.0 + WD=1e-2 (PR #2709) | 0.01323 |
| T_max=10 + gc=1.0 + WD=1e-4 (PR #2680) | 0.01419 |
| T_max=10 + WD=1e-4 (PR #2614) | 0.0197 |

**External target:** 0.0043 (~3x gap remaining)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kohaku:

## Results

**Experiment: T_max=5 + grad-clip=1.0 + WD=1e-2 compound**

| Metric | This run | Baseline (PR #2732) | Delta |
|--------|----------|---------------------|-------|
| **val_primary/surface_mse** | **0.01476** (ep 37) | **0.01271** (ep 40) | +16.1% ❌ |
| test_primary/surface_mse | 0.01738 (ep 41) | — | — |
| full_val/surface_mse_p | 0.0588 (ep 37) | — | — |
| full_val/volume_mse | 0.0874 (ep 37) | — | — |

**Did NOT beat baseline.** WD=1e-2 with T_max=5 is worse than WD=1e-4 (default) with T_max=5.

### Training dynamics

- **Phase transition:** Gradual descent starting ~epoch 18 (0.081 → 0.055 at epoch 21 → 0.016 at epoch 29). No single sharp transition like baseline.
- **Cosine trough alignment:** With T_max=5, troughs at epochs 5, 10, 15, 20, 25, 30, 35, 40. Best epoch 37 is mid-cycle (2 epochs after peak at 36), not at a trough — unlike baseline best at epoch 40 (exact trough).
- **Late-epoch envelope:** Epochs 29-41 oscillate between 0.015-0.050, with the deepest points at ep29 (0.016), ep34 (0.016), ep37 (0.015), ep41 (0.016). The model plateaued — multiple troughs hit ~0.015-0.016 without going deeper.
- **Grad clipping:** 316-358 of 360 batches clipped per epoch (88-99%). Mean grad norm decreased from ~12-23 early to ~10-12 later, indicating stabilization.

### Epoch-by-epoch val_primary/surface_mse (cosine troughs marked with *)

| Epoch | val_primary/surface_mse | Note |
|-------|------------------------|------|
| 5* | 0.2310 | trough |
| 10* | 0.1707 | trough |
| 15* | 0.1729 | trough |
| 20* | 0.0842 | trough |
| 22 | 0.0309 | mid-cycle low |
| 25* | 0.0828 | trough |
| 29 | 0.0160 | mid-cycle low |
| 30* | 0.0206 | trough |
| 34 | 0.0163 | mid-cycle low |
| 35* | 0.0167 | trough |
| **37** | **0.01476** | **BEST (mid-cycle)** |
| 40* | 0.0266 | trough |
| 41 | 0.0157 | final (near peak) |

### What happened

The higher weight decay (1e-2 vs 1e-4 default) appears to over-regularize the model when combined with T_max=5's rapid cycling. The baseline (T_max=5 + WD=1e-4) reached 0.01271 at a clean cosine trough, while this run plateaued at ~0.015-0.016 across multiple cycles without breaking through to deeper basins. The WD=1e-2 + T_max=10 combo (PR #2709, 0.01323) worked better because T_max=10's longer cycles gave the model more optimization time per descent before the LR reset, partially compensating for the stronger regularization.

**Key finding:** T_max=5 and WD=1e-2 are NOT orthogonal improvements. The rapid LR cycling (8 full cycles in 41 epochs) combined with strong regularization creates a competition — the model cannot descend as deeply before the next LR restart disrupts progress. T_max=5 is the dominant lever and prefers lighter regularization.

### Exact command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 \
  --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --epochs 999 --wandb-group airfrans-tmax5-compounds \
  --wandb-name "kohaku/airfrans-tmax5-gc1-wd1e2" --agent kohaku
```

- **W&B run:** i2b7rd02
- **Epochs:** 41 (30-min timeout)

### Suggested follow-ups

1. **T_max=5 + WD=5e-3** — halfway between default 1e-4 and the too-aggressive 1e-2. May find the sweet spot.
2. **T_max=5 + WD=1e-3** — lighter regularization that still adds some benefit over default.
3. **T_max=3 or T_max=4** — even more rapid cycling with default WD to push the T_max axis further.
4. **Longer training budget (180 min)** on the current T_max=5 + WD=1e-4 baseline — since the best is at epoch 40 out of 41, there's likely headroom with more epochs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2741 because ADVISOR: Closing — gc=1.0+T_max=5+WD=1e-2 achieved 0.014756 vs 0.00935 baseline. The gc=1.5 breakthrough makes gc=1.0 experiments dead ends. Will reassign you to a gc=1.5 experiment..
